clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -352,7 +355,9 @@ private TransactionalRequestResult beginCompletingTransaction(TransactionResult
         enqueueRequest(handler);
 
         // If an epoch bump is required for recovery, initialize the transaction after completing the EndTxn request.
-        if (clientSideEpochBumpRequired) {
+        // If we are upgrading to TV2 transactions on the next transaction, also bump the epoch.
+        maybeUpdateTransactionV2Enabled();
```

Comment:
Also transitioning the version here also means that we enable V2 before we fully finish transaction, could you add a comment to the EndTxnHandler.handleResponse that we can receive a V1 response even though we've transitioned into a V2 mode.

Reply:
Also should we remove it from the beginTransaction?  If we somehow call it from there without going through this path, the upgrade flow won't trigger.

Reply:
> Also transitioning the version here also means that we enable V2 before we fully finish transaction, could you add a comment to the EndTxnHandler.handleResponse that we can receive a V1 response even though we've transitioned into a V2 mode.

My understanding is that we do this after we add the EndTxn to the handler. My understanding is that the EndTxn request should always complete before the epoch bump is done (otherwise we would fence ourselves in the epoch bump case) so I thought this was safe, but I can think about this again.

> Also should we remove it from the beginTransaction? If we somehow call it from there without going through this path, the upgrade flow won't trigger.

I have removed it from beginTransaction in this PR already.